File: Important Techniques/floydTortoiseAndHare.c
#include<stdio.h>
#include<stdlib.h>

struct Node
{
    int data;
    struct Node *link;  
};

typedef struct Node Node;
Node *head = NULL;

Node *newNode(int value){
    Node *temp = (Node*)malloc(sizeof(Node));
    if (temp == NULL){
        printf("No memory allocated\n");
    }
    temp->data = value;
    temp->link = NULL;
    return temp;
}

int insertAtBeginSingle(int value){
    if(head == NULL){
        head = newNode(value);
    }
    else{
        Node *temp = newNode(20);
        temp->link = head;
        head = temp;
    }
}

void insertAtBeginCirc(int value){
    Node* temp1 = newNode(value);
    if(head == NULL){
        head = temp1;
        head->link = head;
    }
    else{
        Node* temp = head;
        while(temp->link != head)
        {   
            temp = temp->link;
        }
        temp->link = temp1;
        temp1->link = head;
        head = temp1;
    }
}

int detectCycle(){
    Node* slow = head;
    Node* fast = head;

    while(slow != NULL && fast != NULL && fast->link != NULL) {
        slow = slow->link;
        fast = fast->link->link;
        if (fast == slow) {
            return 1;
        }
    }
    return 0;
}

int main()
{
    // insertAtBeginSingle(2);
    // insertAtBeginSingle(12);
    // insertAtBeginSingle(23);
    // insertAtBeginSingle(21);

    insertAtBeginCirc(3);
    insertAtBeginCirc(33);
    insertAtBeginCirc(34);
    insertAtBeginCirc(30);

    if(detectCycle()){
        printf("Cycle Detected");
    }
    else{
        printf("Cycle Not Detected");
    }
}

File: Important Techniques/floydTortoiseAndHare.py
class Node:
    def __init__(self,value) -> None:
        self.data = value
        self.next = None

class CircularLL:
    def __init__(self) -> None:
        self.head = None

    # circular ll
    # def insertAtBegin(self,value):
    #     temp1 = Node(value)
    #     if not self.head:
    #         self.head = temp1
    #         self.head.next = self.head
    #     else:
    #         temp1.next = self.head
    #         self.head = temp1

    # single ll
    # def addAtBegin(self,value):
    #     if self.head is None:
    #         self.head = Node(value)
    #     else:
    #         temp = Node(value)
    #         temp.next = self.head
    #         self.head = temp

    def detectCycle(self):
        slow = self.head
        fast = self.head
        while(slow is not None and fast is not None and fast.next is not None):
            slow = slow.next
            fast = fast.next.next
         
            if (fast == slow):
                return 1
        return 0
         
    

cll = CircularLL()


# circ ll addtion at beginning
# cll.insertAtBegin(2)
# cll.insertAtBegin(22)
# cll.insertAtBegin(24)
# cll.insertAtBegin(24)

# single ll addtion at beginning
# cll.addAtBegin(2)
# cll.addAtBegin(13)
# cll.addAtBegin(20)
# cll.addAtBegin(22)
print(cll.detectCycle())